Feature/78 database storage

kevin-wijnen opened the pull request.

kevin-wijnen committed:
Database login implementation + Classes migrated to packages

kevin-wijnen committed:
Rename packages to singular

kevin-wijnen committed:
Entity and Repository Relay integrated

kevin-wijnen committed:
Removed unused features & persisting trouble with repository calls

rlemmers committed:
78 ~ Converts ServerSimulator to CommandLineRunner

kevin-wijnen committed:
Added: Relays addition to database when they are not stored

kevin-wijnen committed:
Added: Saving of new Relay status to datebase

This commits completes the feature to store relays in the database

kevin-wijnen committed:
Broken model Relay light status + fix Relay object light status

Added:
- Relay object light status updates by database

Broken:
- Updating the model's Relay light status

kevin-wijnen committed:
Fix: Adding Relays when 'relay' table in database is empty

kevin-wijnen committed:
Fix: Updating existing relays in database

TODO:
- Fix saving new Relay light status to model

kevin-wijnen committed:
Fix: Writing new Relay data to model

Broken: Client GUI of IEC61850bean does not show newest version of
model when booting up, yet to fix. Reloading separate values
show the most recent changes, however.

kevin-wijnen committed:
Schedule table relation added

kevin-wijnen committed:
Schedule database implementation WIP + fix relay model light status

- Schedule implementation work in progress (fix relay search bug
and only save enabled schedules)
- Relay model light status fixed by setting stVal instead of ctlVal

kevin-wijnen committed:
Added enabled schedule saving to database

Yet to implement: Updating schedules in model

kevin-wijnen committed:
General clean up

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/dataclass/Device.java:
```diff
@@ -0,0 +1,144 @@
+        for (int relayNr = 0; relayNr < 4; relayNr++) {
+            logger.info(this.relays[relayNr].toString());
+            for (int scheduleNr = 1; scheduleNr <= 50; scheduleNr++) {
+                logger.info("Schedule {}: {}", scheduleNr, this.getRelay(relayNr + 1).getSchedule(scheduleNr));
```
Only print enabled schedules (to make reading a lot easier).

kevin-wijnen replied:
Done

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/dataclass/Device.java:
```diff
@@ -0,0 +1,144 @@
+                    this.serverWrapper.findModelNode(SWITCH_ROOT + (relayNr + 1) + ".Sche.sche1", Fc.CF),
+                    this.serverWrapper);
+            logger.info("Checking relays in database...");
+            // databaseUtils.checkRelay(this.relays[relayNr]);
```
Remove commented code.

kevin-wijnen replied:
Done

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/dataclass/Device.java:
```diff
@@ -0,0 +1,144 @@
+            logger.info("Set light status for relay {} to value {}", relay, status);
+
+            final BdaBoolean lightStatus = (BdaBoolean) this.serverWrapper
+                    .findModelNode(SWITCH_ROOT + relay + ".Pos.Oper.ctlVal", Fc.CO);
```
This should be Pos.stVal / Fc.ST!

kevin-wijnen replied:
Done

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/dataclass/Schedule.java:
```diff
@@ -0,0 +1,324 @@
+            return this;
+        }
+
+        public ScheduleBuilder isEnabled(boolean isEnabled) {
```
rename the parameter _isEnabled_ to _enabled_.

kevin-wijnen replied:
Done

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/datamodel/ScheduleEntity.java:
```diff
@@ -0,0 +1,155 @@
+
+    }
+
+    public ScheduleEntity(int indexNumber, int day, int timeTypeOn, int timeTypeOff, LocalTime timeOn,
```
Too many parameters. Is this constructor needed?

kevin-wijnen replied:
It is currently used for initializeDatabaseSchedule(), to create a Schedule and send that to the repository for saving it in the database.

kevin-wijnen replied:
Done

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/datarepository/RelayRepository.java:
```diff
@@ -0,0 +1,13 @@
+import com.cgi.iec61850serversimulator.datamodel.RelayEntity;
+
+public interface RelayRepository extends JpaRepository<RelayEntity, Long> {
+    List<RelayEntity> findByIndexNumber(int index_number);
```
Rename _index_number_ to _indexNumber_. Java does not use "snake case" variable names. Also update this for other variable names in you code which have snake casing.

kevin-wijnen replied:
Done

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/functionclass/ActionExecutor.java:
```diff
@@ -0,0 +1,58 @@
+    private static final Logger logger = LoggerFactory.getLogger(ServerSimulator.class);
+    private static final String PRINT_SERVER_MODEL_KEY = "p";
+    private static final String DEVICE_SHOW_MODEL = "d";
+    public ServerSap serverSap = null;
+    public ServerModel serverModel = null;
+    public Device device = null;
```
- Make serverSap and device private.
- Remote the serverModel field.

kevin-wijnen replied:
Done

kevin-wijnen committed:
Fix: DatabaseUtils in Device class

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/functionclass/DatabaseUtils.java:
```diff
@@ -0,0 +1,176 @@
+        logger.info("Updating model relay  database.");
+        int relayNr = relay.getIndexNumber();
+        List<RelayEntity> databaseRelay = this.relayRepository.findByIndexNumber(relayNr);
+        final BdaBoolean modelRelayLightStatus = (BdaBoolean) this.serverWrapper
+                .findModelNode(SWITCH_ROOT + relayNr + ".Pos.stVal", Fc.ST);
+        modelRelayLightStatus.setValue(databaseRelay.get(0).isLightStatus());
+        List<BasicDataAttribute> bdas = new ArrayList<BasicDataAttribute>();
+        bdas.add(modelRelayLightStatus);
+        relay.setLight(databaseRelay.get(0).isLightStatus());
+        this.serverWrapper.setValues(bdas);
```
The Device class also contains code to set the light status. Ensure there's no duplication.

kevin-wijnen replied:
Done, by removing it from the Device class. Only the Relay class' light status function has been called so far, prompting choosing the Relay class' implementation over the Device's implementation.

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/functionclass/EventDataListener.java:
```diff
@@ -107,7 +105,7 @@ public EventDataListener(final Device device, Scheduler scheduler) {
                     // Ruud: example to modify a light status, which is what should
                     // happen at the point in
                     // time when a schedule moment should change a light status.
-                    this.device.setLightStatus(2, newValue);
+                    // this.device.setLightStatus(2, newValue);
```
Remove this old example code/comment.

kevin-wijnen replied:
Done

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/functionclass/ServerSimulator.java:
```diff
@@ -92,22 +113,38 @@ public void run() {
+        DatabaseUtils databaseUtils = new DatabaseUtils(serverWrapper);
+        databaseUtils.setRelayRepository(this.relayRepository);
+        databaseUtils.setScheduleRepository(this.scheduleRepository);
+        for (int i = 1; i < device.getRelays().length + 1; i++) {
```
Cleaner: "<= device.getRelays().length" instead of "< device.getRelays().length + 1"
More meaningful: "relayNumber" instead of i.

kevin-wijnen replied:
Done

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/functionclass/ServerSimulator.java:
```diff
@@ -92,22 +113,38 @@ public void run() {
+            Relay relay = device.getRelay(i);
+            databaseUtils.checkRelay(relay);
+
+            for (int j = 1; j < 51; j++) {
```
Cleaner: "<= 50" instead of "<51"
More meaningful: "scheduleNumber" instead of j.

kevin-wijnen replied:
Done

kevin-wijnen committed:
Model schedule updating with database data implemented

Yet to implement: EventDataListener updating database schedules

kevin-wijnen committed:
Feedback integrated & integration Schedule persistent storage

Despite the implementations of schedule saving, it seems to not function
yet? Bug should get investigated into!

kevin-wijnen commented:
Important to note (as said in commit message): Schedule storage implementation has been worked on, but besides initialization, storing new schedules does not seem to function properly. Need to investigate the bug to correct its behavior.

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/functionclass/DatabaseUtils.java:
```diff
@@ -162,15 +221,24 @@ public void initializeDatabaseSchedule(Schedule schedule, int relayNr) {
+        int index_number = schedule.getIndexNumber();
+        int relayNr = schedule.getRelayNr();
+
+        List<ScheduleEntity> scheduleEntities = this.scheduleRepository.findByIndexNumberAndRelayId(index_number,
```
index_number => indexNumber

kevin-wijnen replied:
Done

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/functionclass/DatabaseUtils.java:
```diff
@@ -162,15 +221,24 @@ public void initializeDatabaseSchedule(Schedule schedule, int relayNr) {
-    public void updateDatabaseSchedule(Relay relay) {
-        // Updating database relay
+    public void updateDatabaseSchedule(Schedule schedule) {
+        // Updating database schedule
```
This comment can be removed. The method name already shows it's updating a schedule in the database.

kevin-wijnen replied:
Done, same for updateDatabaseRelay.

kevin-wijnen committed:
Removed unnecessary schedule reports (above 50)

kevin-wijnen committed:
Multiple schedule database fixes & shutdown bug fixed

Fix: Creation of duplicate schedules when building the schedule table
Fix: Saving of schedules by proper id works now
Fix: Not creating new disabled schedules in database anymore
Fix: Shutting down the program by pressing Q is fixed (scheduler issue)
Layout: Renamed initializeDatabase... functions to createDatabase...

kevin-wijnen committed:
ScheduleEntity has a builder now instead of just a regular constructor

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/functionclass/DatabaseUtils.java:
```diff
@@ -107,16 +112,20 @@ public void checkSchedule(Schedule schedule, int relayNr) {
+            this.updateModelSchedule(schedule, relayNr);
         } else {
 
             if (schedule.isEnabled()) {
```
Why isn't it necessary to check isEnabled for entries that were found in the DB?

kevin-wijnen replied:
Because deletion was not possible at that time. This has now been implemented along with a deletion feature to delete disabled schedules in the DB to save on space.

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/functionclass/DatabaseUtils.java:
```diff
@@ -157,20 +226,54 @@ public void initializeDatabaseSchedule(Schedule schedule, int relayNr) {
+        scheduleBuilder.setEnabled(enabled);
+        ScheduleEntity toUpdateSchedule = scheduleBuilder.buildScheduleEntity();
+
+        /*
+         * ScheduleEntity toUpdateSchedule = new ScheduleEntity(indexNumber, day,
+         * timeOnType, timeOffType, timeOn, timeOff, burningMinutes, enabled,
+         * relayEntity);
+         */
```
Remove this commented code.

kevin-wijnen replied:
Done.

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/datamodel/ScheduleEntity.java:
```diff
@@ -15,6 +15,75 @@
+            this.relayEntity = relayEntity;
+        }
+
+        public void setIndexNumber(int indexNumber) {
+            this.indexNumber = indexNumber;
+        }
```
This method should be deleted, because the indexNumber is already set in the constructor of this builder.

kevin-wijnen replied:
Done.

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/datamodel/ScheduleEntity.java:
```diff
@@ -15,6 +15,75 @@
+        private RelayEntity relayEntity;
+
+        public ScheduleEntityBuilder(int scheduleNr, RelayEntity relayEntity) {
+            this.indexNumber = scheduleNr;
```
"scheduleNr" as parameter name and "indexNumber" as instance variable? They should be using the same name, otherwise it's confusing.

kevin-wijnen replied:
Done.

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/datamodel/ScheduleEntity.java:
```diff
@@ -15,6 +15,75 @@
+            this.indexNumber = indexNumber;
+        }
+
+        public void setDay(int day) {
```
- A builder should use either "withDay" or just "day" as method name.
- The method should return the builder, so you can chain calls to the builder.

kevin-wijnen replied:
Done, corrected the entire builder.

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/datamodel/ScheduleEntity.java:
```diff
@@ -15,6 +15,75 @@
 @Table(name = "schedule")
 public class ScheduleEntity {
 
+    public static class ScheduleEntityBuilder {
```
Normally builders are put at the bottom of a class.

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/functionclass/DatabaseUtils.java:
```diff
@@ -157,20 +226,54 @@ public void initializeDatabaseSchedule(Schedule schedule, int relayNr) {
+
+        // Builder
+        ScheduleEntityBuilder scheduleBuilder = new ScheduleEntityBuilder(indexNumber, relayEntity);
+        scheduleBuilder.setDay(day);
+        scheduleBuilder.setTimeTypeOn(timeOnType);
+        scheduleBuilder.setTimeTypeOff(timeOffType);
+        scheduleBuilder.setTimeOn(timeOn);
+        scheduleBuilder.setTimeOff(timeOff);
+        scheduleBuilder.setBurningMinutes(burningMinutes);
+        scheduleBuilder.setEnabled(enabled);
+        ScheduleEntity toUpdateSchedule = scheduleBuilder.buildScheduleEntity();
```
This will change, when the comments about the builder have been processed.

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/functionclass/EventDataListener.java:
```diff
@@ -53,10 +53,9 @@ public EventDataListener(final Device device, Scheduler scheduler, DatabaseUtils
-                logger.info("Bda Type: " + bda.getBasicType());
-                // System.out.println("Relay array created." +
-                // relayScheduleNumbers);
+                // logger.info(dataAttribute);
+                // logger.info("Bda Type: " + bda.getBasicType());
```
Remove this commented code.

kevin-wijnen replied:
Done.

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/functionclass/EventDataListener.java:
```diff
@@ -154,171 +148,241 @@ public EventDataListener(final Device device, Scheduler scheduler, DatabaseUtils
-                    } catch (final Exception e) {
-                        logger.info("Schedules above 50 are not implemented in the GXF platform. Skip ...");
+
+                    if (scheduleIndex <= 50) {
```
This check is copied a number of times. Other parts of the case block also look a lot alike. Check how you can eliminate all this duplication.

kevin-wijnen replied:
Done.

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/functionclass/EventDataListener.java:
```diff
@@ -154,171 +148,241 @@ public EventDataListener(final Device device, Scheduler scheduler, DatabaseUtils
                 default:
                     // When BDA is not used in a class
-                    logger.info("Unimplemented value found, " + "'" + dataAttribute + "'" + ", skipped.");
+                    // logger.info("Unimplemented value found, " + "'" + dataAttribute + "'" + ",
+                    // skipped.");
```
Remove this commented code.

kevin-wijnen replied:
Done.

kevin-wijnen committed:
Small feedback-based layout changes

kevin-wijnen committed:
Builder feedback processed

kevin-wijnen committed:
EventDataListener rework

Layout: Removed duplicate code and unnecessary logging
Layout: Changed conditional checks to be more global

kevin-wijnen committed:
Deletion of disabled schedules

Feature: Deleting disabled schedules to save space in database
Note about feature: The enabled schedule in the model is created
constantly. Need to find solution for it.

kevin-wijnen committed:
Tasks will not be created when relative time < 0

Fix: Runnables with the relative time of 0 or lower will not be made
when calculating switching moments.
